### DISCUSSION TITLE
Using internal structures inside SLANG?

### DISCUSSION THREAD

**Discussion by Practical-UVM-Step-By-Step** [MikePopoloski/slang#709] [General]
Hi Mike:
I had two questions.

I am wondering if there is an API that will expose the equivalent of the contents of ParserMetaData struct
Out to the user or whether there is an easy way to get that kind of structured information from the compilation itself.
If there is one, is there some test I can look for an example? 

Q1: Do you recommend reaching out into it or do you recommend a new visitor pattern to essentially repeat what you already did and maybe build a separate data structure?

Q2: 
	Consider the example below

	class myclass extends uvm_object;
		int a;
		string mister;
	`uvm_component_utils_begin(my class)
		`uvm_field_int(a,UVM_ALL_ON)
		`uvm_field_string(mister, UVM_ALL_ON|UVM_NOCOPY)
	`uvm_component_utils_end

	task configureme()
	`uvm_info(“CONFIGURING”,”test of Configuration”,UVM_LOW)
	..
	end task

	// Other usual class stuff
	endclass

I’m trying to essentially get the arguments, counts etc of the `uvm_field_int macros and the other `uvm_macros but they would be “pre-processed” and the documentation says that the contents of the `uvm macros would be placed in separate buffers and to look at the buffer_ids. 

Would you be able to share some pseudo code/ideas of how you envision accomplishing this?
Thanks!

**Comment by MikePopoloski** [comment]
You can use the existing ParserMetadata if you want, it's accessible via the API. I'm not sure it contains what you want to know though.

I'm not sure I understood your example; are you saying you want to know how many times the macros are invoked? Preprocessing happens pre-parser so you'd need to work at a pretty low level of the API to get that information.
